gcsio/src/main/java/com/google/cloud/hadoop/gcsio/AdaptiveFileAccessPattern.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.gcsio;
+
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageReadOptions.Fadvise;
+import com.google.common.flogger.GoogleLogger;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.ListIterator;
+
+class AdaptiveFileAccessPattern implements Closeable {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private final StorageResourceId resourceId;
+  private final GoogleCloudStorageReadOptions readOptions;
+  private boolean isPatternOverriden = false;
+  private boolean randomAccess;
+  private long lastServedIndex = -1;
+  // Keeps track of distance between consecutive requests
+  private BoundedList<Long> consecutiveRequestsDistances;
+
+  @Override
+  public void close() throws IOException {
+    if (consecutiveRequestsDistances != null) {
+      consecutiveRequestsDistances = null;
+    }
+  }
+
+  class BoundedList<E> extends LinkedList<E> {
+    private int limit;
+
+    public BoundedList(int limit) {
+      this.limit = limit;
+    }
+
+    @Override
+    public boolean add(E o) {
+      super.add(o);
+      while (size() > limit) {
+        super.removeFirst();
+      }
+      return true;
+    }
+  }
+
+  public AdaptiveFileAccessPattern(
+      StorageResourceId resourceId, GoogleCloudStorageReadOptions readOptions) {
+    this.resourceId = resourceId;
+    this.readOptions = readOptions;
+    this.randomAccess =
+        readOptions.getFadvise() == Fadvise.AUTO_RANDOM
+            || readOptions.getFadvise() == Fadvise.RANDOM;
+    if (readOptions.getFadvise() == Fadvise.AUTO_RANDOM) {
+      consecutiveRequestsDistances = new BoundedList<>(readOptions.getFadviseRequestTrackCount());
+    }
+  }
+
+  public void updateLastServedIndex(long position) {
+    this.lastServedIndex = position;
+  }
+
+  public boolean isRandomAccessPattern() {
+    return randomAccess;
+  }
+
+  public void updateAccessPattern(long currentPosition) {
+    if (isPatternOverriden) {
+      logger.atFiner().log(
+          "Will bypass computing access pattern as it's overriden for resource :%s", resourceId);
+      return;
+    }
+    if (readOptions.getFadvise() == Fadvise.AUTO_RANDOM) {
+      if (isSequentialAccessPattern(currentPosition)) {
```

Comment:
will it only switch from sequential -> random once and not go back to sequential? Is this intentional?

Reply:
Yeah, it is as per the request as mentioned in the document AUTO_RANDOM starts assuming "RANDOM" file access and adapts to "SEQUENTIAL" if read pattern is so. Once it's adapted it will not flip again.